extensions/standard-processors/processors/PutTCP.cpp

```diff
@@ -160,339 +177,145 @@ void PutTCP::onSchedule(core::ProcessContext* const context, core::ProcessSessio
-  auto ssl_context = getSslContext(ssl_context_service_);
-  auto socket = std::make_shared<SslSocket>(io_context_, ssl_context);
-  startConnect(resolved_query.begin(), socket);
-  deadline_.async_wait([&](std::error_code error_code) -> void {
-    checkDeadline(error_code, socket.get());
-  });
-  io_context_.run();
-  if (last_error_)
-    return nonstd::make_unexpected(last_error_);
-  return socket;
+template<class SocketType>
+asio::awaitable<std::error_code> ConnectionHandler<SocketType>::sendStreamWithDelimiter(const std::shared_ptr<io::InputStream>& stream_to_send,
+                                                                                        const std::vector<std::byte>& delimiter,
+                                                                                        asio::io_context& io_context) {
+  if (auto connection_error = co_await setupUsableSocket(io_context))  // NOLINT
+    co_return connection_error;
+  co_return co_await send(stream_to_send, delimiter);
 }
 
 template<class SocketType>
-nonstd::expected<void, std::error_code> ConnectionHandler<SocketType>::sendDataToSocket(const std::shared_ptr<SocketType>& socket,
-                                                                                        const std::shared_ptr<io::InputStream>& flow_file_content_stream,
-                                                                                        const std::vector<std::byte>& delimiter) {
-  if (!socket || !socket->lowest_layer().is_open())
-    return nonstd::make_unexpected(asio::error::not_socket);
-
-  deadline_.expires_after(timeout_);
-  deadline_.async_wait([&](std::error_code error_code) -> void {
-    checkDeadline(error_code, socket.get());
-  });
-  io_context_.restart();
+asio::awaitable<std::error_code> ConnectionHandler<SocketType>::send(const std::shared_ptr<io::InputStream>& stream_to_send,
+                                                                     const std::vector<std::byte>& delimiter) {
+  gsl_Expects(hasUsableSocket());
 
   std::vector<std::byte> data_chunk;
   data_chunk.resize(chunk_size);
-
   gsl::span<std::byte> buffer{data_chunk};
-  size_t num_read = flow_file_content_stream->read(buffer);
-  logger_->log_trace("read %zu bytes from flowfile", num_read);
-  asio::async_write(*socket, asio::buffer(data_chunk, num_read), [&](const std::error_code err, std::size_t bytes_written) {
-    handleWrite(err, bytes_written, flow_file_content_stream, delimiter, socket);
-  });
-  deadline_.async_wait([&](std::error_code error_code) -> void {
-    checkDeadline(error_code, socket.get());
-  });
-  io_context_.run();
-  if (last_error_)
-    return nonstd::make_unexpected(last_error_);
-  last_used_ = std::chrono::steady_clock::now();
-  return {};
-}
+  while (stream_to_send->tell() < stream_to_send->size()) {
+    size_t num_read = stream_to_send->read(buffer);
```

Comment:
should we handle read error here?

Reply:
good idea :+1: https://github.com/apache/nifi-minifi-cpp/pull/1457/commits/70d6a798090869c4c01b2dd7dc84eeaf9dcef2a3#diff-a62ddb0368e1813dff2a90fe7433e148bfc7b21b7d7ad9c8a066fd040d209ac9R307-R308

Reply:
`io::isError` seems to be used instead of direct comparison with the error value

Reply:
I've changed it in https://github.com/apache/nifi-minifi-cpp/pull/1457/commits/a76b49d2389cc2c51e1eb3aad513f9e638b44301